### PR TITLE
Code fixes for correctness and readability

### PR DESCRIPTION
**Pull request by ivanjayapurna**
- fixed bug in get_rep
- added comma in README
- added CHANGELOG 
- ran make format & style

### CODE REVIEW

**Comment by ericmjl** [comment]
Wonderful, thanks @ivanjayapurna!

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Changelog
+
+In the changelog, @ElArkk and @ericmjl would like to acknowledge contributors who have helped us with anything on the project, big or small.
+
+<Please add your contribution to the top>
+
+- 30 March 2019: Code fixes for correctness and readability by @ivanjayapurna.
```
Comment (ericmjl) [inline CHANGELOG.md:7]:
```suggestion
- 30 March 2019: Code fixes for correctness and readability, and a manual evotuning function by @ivanjayapurna,
```

File: jax_unirep/evotuning.py [hunk 2]
```diff
@@ -411,3 +410,19 @@ def evotune(
...
+
+
+def evotune_manual(
+    sequences: List[str], n_epochs: int = 65, learning_rate: float = 0.00001
+):
+    """
+    Evolutionarily tune the model to a set of sequences,
+    Same as evotune() function but manually selecting epochs and learning rate,
+    instead of using Optuna.
+    """
```
Comment (ericmjl) [inline jax_unirep/evotuning.py:422]:
```suggestion
    :param sequences: The set of sequences to tune to.
    :param n_epochs: The number of epochs to fit.
    :param learning_rate: Optimizer learning rate.
    """
```

File: jax_unirep/evotuning.py [hunk 3]
```diff
@@ -196,6 +194,118 @@ def step(i, state):
...
+        if (i + 1) % steps_per_print == 0:
+
+            # calculate loss for in & out domain validation sets
+            # then print it.
+            # TODO: print to .log file??
+
+            print(
+                "Iteration {0}: in-val-loss={1}, out-val-loss:{2}".format(
+                    i,
+                    avg_loss(in_val_seqs, params),
```
Comment (ElArkk) [inline jax_unirep/evotuning.py:299]: For consistency, I think it would be better to use f-string syntax here.

File: jax_unirep/evotuning.py [hunk 4]
```diff
@@ -196,6 +194,118 @@ def step(i, state):
...
+    """
+    Return average loss of a set of parameters,
+    on a set of sequences.
+
+    :param sequences: sequences (in standard AA format)
+    :param params: parameters (i.e. from training)
+    """
+
+    sequences = onp.array(sequences)
+
```
Comment (ElArkk) [inline jax_unirep/evotuning.py:207]:
```suggestion
```

File: jax_unirep/evotuning.py (hunk 3 above)
Comment (ericmjl) [reply jax_unirep/evotuning.py:296]: @ElArkk & @ivanjayapurna - suggestion below. 

Not sure if it's possible to directly commit from PR page, though worth trying.

```suggestion
            in_val_loss, out_val_loss = avg_loss(in_val_seqs, params)
            print(
                f"Iteration {i}: in-val-loss={in_val_loss}, out-val-loss:{out_val_loss}"
```

File: jax_unirep/evotuning.py [hunk 5]
```diff
@@ -196,6 +194,115 @@ def step(i, state):
...
+            # calculate loss for in & out domain validation sets
+            # then print it.
+            # TODO: print to .log file??
+
+            print(
+                "Iteration {0}: in-val-loss={1}, out-val-loss:{2}".format(
+                    i,
+                    avg_loss(in_val_seqs, params),
+                    avg_loss(out_val_seqs, params),
```
Comment (ericmjl) [inline jax_unirep/evotuning.py:297]:
```suggestion
            print(
                f"Iteration {i}: in-val-loss={avg_loss(in_val_seqs, params)}, out-val-loss:{avg_loss(out_val_seqs, params)}"
```

File: jax_unirep/evotuning.py [hunk 6]
```diff
@@ -196,6 +194,114 @@ def step(i, state):
...
+            # calculate loss for in & out domain validation sets
+            # then print it.
+            # TODO: print to .log file??
+
+            print(
+                f"Iteration {i}: "
+                f"in-val-loss={avg_loss(in_val_seqs, params)}, "
+                f"out-val-loss:{avg_loss(out_val_seqs, params)}"
+                )
+            )
```
Comment (ericmjl) [inline jax_unirep/evotuning.py:298]:
```suggestion
            print(
                f"Iteration {i}: "
                + f"in-val-loss={avg_loss(in_val_seqs, params)}, "
                + f"out-val-loss:{avg_loss(out_val_seqs, params)}"
            )
```

File: jax_unirep/evotuning.py [hunk 7]
```diff
@@ -196,6 +195,126 @@ def step(i, state):
...
+    xs, ys = length_batch_input_outputs(sequences)
+
+    sum_loss = 0
+    for x, y in zip(xs, ys):
+        sum_loss += evotune_loss(params, inputs=x, targets=y) * len(x)
+
+    return sum_loss / len(sequences)
+
+
+def fit_manual(
```
Comment (ericmjl) [inline jax_unirep/evotuning.py:215]: Some issues I'm seeing, now that I see the extra commits coming in. Firstly, it seems to me that there's `fit_manual` has a lot copied from `fit`, but tweaked slightly. This isn't the best coding practice for a software library. 

Secondly, the API has extra pieces compared to `fit`, making the API inconsistent. In particular, the `in_val_seqs`, `out_val_seqs`, `proj_name`, and `steps_per_print` kwargs don't have semantic counterparts in `evotune()`.

Finally, `fit_manual` doesn't do one "unit" thing - it fits, and also handles data I/O (to disk). Generally, I have seen libraries leave data I/O to a separate function that the user has to explicitly call, rather than embed it inside a function. In a "library" function, it's preferable for functions to be "unit".

File: jax_unirep/evotuning.py [hunk 8]
```diff
@@ -196,6 +195,126 @@ def step(i, state):
     return get_params(state)
 
 
+def avg_loss(sequences, params):
```
Comment (ericmjl) [inline jax_unirep/evotuning.py:198]: Let's get in a software test for this function.

**Comment by ivanjayapurna** [comment]
> I'd actually prefer to see `fit_manual` and `evotune_manual` as an example that one could copy/paste as a script, rather than a function to call from the library. My proposal would be to put it in `examples/manual_evotuning.py`. `evotune()` was provided as a high-level convenience function for someone who doesn't want to bother with the process of tuning; if there's a "manual" step, then it means direct user intervention is at play here, which means the code is best shown as an example instead.
> 
> What do you think? Btw, please feel free to push back if you think differently!

Makes sense, I agree. Sorry I actually didn't intend on all the code edits I was doing to be part of this PR (originally just wanted to fix the first bug that stopped get_rep from running and push that) - evotuning with manual parameters I implemented more for myself because thats what I want to use this library to do!

**Comment by ElArkk** [comment]
@ivanjayapurna @ericmjl since `get_rep` is broken right now on `master`, how about splitting the work here into two PR's? The first one only containing the fix and changelog, so we can merge that one ASAP. 

The manual evotuning part has some nice features to it, such as the ability to supply a validation set. 
But, if someone wants to fine-tune the model for a fixed number of epochs with a fixed learning rate, that could also be done by fixing the `low` and `high` values of `n_epochs_config` and `lr_config` to be the same, and setting `n_trials` to 1 in `evotune`. Maybe we should make this more clear in the readme.

I agree with @ericmjl that we should stick to just having `evotune` in the library.

**Comment by ivanjayapurna** [comment]
> @ivanjayapurna @ericmjl since `get_rep` is broken right now on `master`, how about splitting the work here into two PR's? The first one only containing the fix and changelog, so we can merge that one ASAP.
> 
> The manual evotuning part has some nice features to it, such as the ability to supply a validation set.
> But, if someone wants to fine-tune the model for a fixed number of epochs with a fixed learning rate, that could also be done by fixing the `low` and `high` values of `n_epochs_config` and `learning_rate_config` to be the same, and setting `n_trials` to 1 in `evotune`. Maybe we should make this more clear in the readme.
> 
> I agree that we should stick to just having `evotune` in the library.

I agree, lets split this into 2 PRs. Also what are your thoughts then of only keeping evotune but just adding the ability to optionally supply validation sets? We can also keep the ability to dump paramaters as a separate function.

**Comment by ericmjl** [comment]
@ivanjayapurna @ElArkk great discussion thus far.

Let's take some action at this point.

Firstly, @ivanjayapurna, can you shrink down the PR to just the original four bullet points you put up? That way, can get them in. @ElArkk, in the interest of helping you get comfortable with taking these decisions, I think you should take the decision to "hit dat green button".

I'm not sure how familiar you are with `git`, @ivanjayapurna, but a safe prior assumption is that most who submit PRs are good enough, so you should know how to save the rest of the good work you've done here. In case you aren't, simply cut out what's not relevant into a separate file outside of your local repo clone, and push. Let us merge, delete your fork, and then we can go again with a different PR for the other things.

The reason I suggest deleting your fork is because you had submitted a PR from `master`. From my prior experience with contribution to `matplotlib` and `bokeh`, this isn't an idiomatic workflow; rather, once you fork, you should also create a branch on your fork, from which you PR back in. It's a bit long to explain here why this has become an idiomatic way of working, but the short answer is that it keeps the commit histories on forks clean. (GitHub also nudges us to delete forks once we're done PR-ing... but I think there's a different motivation there. We can chat about this on end.)

Once you've gotten this PR merged in, please go ahead and open up a second PR for manual evotuning.

-----

On the nomenclature between "test" and "validation" set, I only know that I've been supremely confused by the terminology as it's never been clarified. On the other hand, the use of the term "holdout", which implies that the data were held out from the modelling workflow, and is only used for evaluation, is extremely clarifying. 

My suggestion here is as follows: pick one that follows the conventions in jax-unirep, and then thoroughly document it in the docstrings, with emphasis on "holdout" being present there so we can be very clear on what is used for final reporting of loss.

@ivanjayapurna, how do you feel about this? Will it be too much to handle, or too troublesome for you? Let us know, always happy to help, and I'm open to doing a pair coding call with you if needed. Just let us know.

**Comment by ivanjayapurna** [comment]
Sounds good, I'll revert the PR back to the original 4 bullets today so @ElArkk can confirm it!

Later this week I'll then open a 2nd PR to discuss changes to evotuning (I think the ability to manually supply an out of domain holdout set is really helpful). Additionally, I think the API for supplying different weights to get_rep() could do with further clarification and thats something else we could discuss. Alternatively, I'm also happy to not merge my evotuning changes at all (or just include it as an example use-case script). I really just wrote that code as that's how I personally wanted to use this library - sorry it all got included in this PR!

File: jax_unirep/evotuning.py [hunk 9]
```diff
@@ -4,23 +4,22 @@
...
 from jax_unirep.losses import _neg_cross_entropy_loss
+from sklearn.model_selection import KFold, train_test_split
 
 from .layers import mLSTM1900, mLSTM1900_AvgHidden, mLSTM1900_HiddenStates
 from .losses import neg_cross_entropy_loss
 from .params import add_dense_params
 from .utils import (
     aa_seq_to_int,
     batch_sequences,
+    dump_params,
```
Comment (ElArkk) [inline jax_unirep/evotuning.py:22]:
```suggestion
```

**Comment by ElArkk** [comment]
@ivanjayapurna thank you for cleaning things up! With all tests passing, I'm going to hit merge on this one.

**Comment by ericmjl** [comment]
Woohoo! Thank you @ivanjayapurna! :tada:
